Move to jstransformer

ismay opened the pull request:
~~Still todo:~~

~~\- [x] Update travis (test on multiple node versions)~~
~~\- [x] Remove hound~~
~~\- [x] Add debug messsages~~
~~\- [x] Stop processing at the filename (instead of relying on there being no transform)~~
~~\- [x] Currently uses private metalsmith properties~~
~~\- [x] Fix package.json~~
~~\- [x] Check if file has an extension is superfluous, but might result in a minor speed increase~~
~~\- [x] Test for common problems with templating languages (jade, haml, nunjucks)~~
~~\- [x] Currently removes last extension after every transform (should use outputformat, otherwise it'll transform `index.md` to `index`) (see [here](https://github.com/jstransformers/jstransformer#outputformat))~~
~~\- [ ] Make nunjucks test work cross-directory~~
~~\- [ ] Make haml test work~~

~~Update readme & history:~~

~~\- [x] credit Rob Loach for the inspiration~~
~~\- [ ] mention node support (4,5,6)~~
~~\- [ ] mention debugging option~~
~~\- [x] documents opts~~
~~\- [ ] documents options passthrough (mention options.filename) and the workings of this with jstransformers (partials etc.)~~
~~\- [x] mention that transformers need to be installed~~

~~Optimisations~~

~~\- [ ] Performance check (move to async)~~
~~\- [ ] Move scanning for partials to external plugin (add partials to metadata)~~

See below for updated todo

Commit:
Move to jstransformer

Commit:
Add travis

Commit:
Skip files without an extension

Commit:
Add debug

Commit:
Stop rendering at filename

Commit:
Update package.json

Commit:
Improve debug messages

Commit:
Don't use private metalsmith properties

Commit:
Update tests

Commit:
Modularize

Commit:
Replace last extension when transformed

Commit:
Remove duplicate content saving

Commit:
Add lang tests

Commit:
Remove build dirs from version control

Commit:
Update gitignore

doodzik commented:
This seems like an enhance to what we had previously.
It is good that you moved parts of the functionality into their own functions.
I would try to simplify `render-file.js` more.

ismay commented:
Yeah I tried to keep it compartmentalised.

> I would try to simplify render-file.js more.

Definitely, I'm gonna see what I can simplify further.

woodyrew commented:
An interesting PR. What was the motivation to move away from consolidate?

Are you considering having a common core shared by `metalsmith-in-place` and `metalsmith-layouts`?

ismay commented:
@woodyrew 

> What was the motivation to move away from consolidate?

Modularity mostly. The fact that transformers can be easily and independently updated will hopefully mean that issues with rendering engines can be fixed more easily than with consolidate. Also, so far it seems that the jstransformers project is more active. However, we'll be testing things out first with a beta. 

> Are you considering having a common core shared by metalsmith-in-place and metalsmith-layouts?

Definitely! That's actually the next thing on our todolist.

Commit:
Remove nunjucks test

Commit:
Add readme and history

ismay commented:
@doodzik I consider this PR ready for the beta release, do you want to give it a final check? If you agree we can squash and merge.

After that I'll release it on npm and we can fix all the other stuff like abstracting etc. (because there's still quite some things I'd like to change an improve on).

ismay commented:
Stuff to put in the issues after releasing:

#### Large changes

- rename `layouts` and `in-place` as the names are a bit confusing. Make it more obvious that `in-place` is the main lib, and that `layouts` is just a polyfill for languages without extends (so make `in-place` the main lib and deprecate `layouts`. But this is a major change so we should be careful with this.
- release scanning for partials, helpers, etc. as separate libs.
- design for putting everything in `src`, instead of in a separate folder.
- abstract rendering to separate lib so that the apis of `in-place` and `layouts` are more agnostic to the rendering lib being used. Allows it to be overridden easily and simplifies testing.

#### Language specific tests

Also depends on the changes we want to make to our testing methodology

- make nunjucks test work cross-directory
- make haml test work

#### Update readme

- mention node support (4,5,6)
- mention debugging option (like npm run debug)
- documents options passthrough, options.filename and the workings of this with jstransformers (for partials, helpers etc.)

#### Optimisations

- performance check (async)

doodzik commented:
I'm going to look at it today in the evening or tomorrow :+:

woodyrew commented:
What name changes did you have in mind regarding naming? I agree there is confusion regarding the names and the behaviour of the two packages. `metalsmith-templates` was split into `metalsmith-layouts` and `metalsmith-in-place` so they're more focused in their task as far as I am aware. I'm not saying they shouldn't be renamed or int the case of `layouts` be deprecated but as **heavily** relied upon plugin I think it should be greatly considered.

On a side note, do you know if the switch to `jstransformer` along with the other changes will provide a performance increase to layouts? On all my projects, this is the plugin that takes the longest time (using Handlebars), often much longer than all other plugins combined.

I hope I'm not treading on any toes, my aim is to help. You're both doing excellent work and it's greatly appreciated :smile:

ismay commented:
> I hope I'm not treading on any toes, my aim is to help.

Oh no, definitely not, I really appreciate the feedback 👍 😄  

>I'm not saying they shouldn't be renamed or int the case of layouts be deprecated but as heavily relied upon plugin I think it should be greatly considered.

Completely agree, if we're going to change this, it should be considered **very** carefully. I didn't really formulate it well because @doodzik and I already discussed it and this was just a reference to that discussion. 

Basically we just thought that it would be good for people to realise that in-place is what you should use in 99% of the cases. It simplifies everything because you can just use the functionality of whichever templating language you use. If you use a complete templating language you can use extends and includes without a problem.

The only reason I see for using layouts is if you absolutely _have_ to use a templating language that doesn't support extends natively (like handlebars). So a polyfill for extends.

Now unfortunately layouts gets a lot more use, and it is probably one of the the first plugins that newcomers to metalsmith use. Which complicates the learning process for newcomers because it does some counterintuitive stuff (like not process syntax in the source file). We get a lot of issues in the layouts repo that are basically the result of people running into trouble with layouts, where they'd probably be fine with in-place.

So basically we're just thinking about how we can help people understand the roles of both plugins, and a name change was just one of the suggested options (because in-place is not as descriptive to newcomers).

> On a side note, do you know if the switch to jstransformer along with the other changes will provide a performance increase to layouts

Not yet. We were planning on doing perf testing and optimisation after this pr, and it was definitely something that we were also thinking about. Especially with a large collection of files this plugin can take a long time and it would be nice to get that down.

doodzik commented on index.js:
```diff
@@ -0,0 +1,40 @@
+    // Loop through all files
+    Object.keys(files).forEach((filename) => {
+      // Check file, skip processing if it fails
+      if (!checkFile(filename, pattern)) return done();
```
If we call `done` here isn't it ending the whole plugin from executing?

woodyrew replied:
Ah, you're using es6 anyway.

doodzik commented on index.js:
```diff
@@ -0,0 +1,40 @@
+      if (!checkFile(filename, pattern)) return done();
+
+      // Transform file
+      const newname = renderFile(files, filename, metadata, source, options);
```
It would be better if we aren't modifying files from the render function as it adds unnecessary complexion.
But I think we can change that in another pr.

doodzik commented on lib/check-file.js:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  // Only process files with an extension
+  if (!contains(filename, '.')) {
```
we can use [includes](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/includes) for that purpose

doodzik committed:
make if statement more readable

doodzik commented:
This seems good. Before merging it would be good to verify if the `done` call behaves as expected or if we need to alter it.

woodyrew commented:
@ismay @doodzik Added a line comment regarding the `done()`

ismay commented:
@woodyrew I might be missing it, but I don't see your line comment. Did it get posted?

@doodzik You might be right about the `done()`. Couldn't find documentation on the recommended usage of the done() callback with a quick search, but I'd also expect it to be called only when all processing is done (even though tests don't seem to misbehave when checking with an `npm run debug`, but that might be because the testing isn't done in conjunction with other plugins that could interfere). I agree with your other suggestions, sounds good 👍

woodyrew commented:
@ismay Yes, it was _pending_ in GH's new feature. https://github.com/superwolff/metalsmith-in-place/pull/52#r78805065

ismay commented:
Ah. So the `done()` will have to be changed before merging then, my bad. I'll fix it and implement the other suggestions as well, then we should be ready to go.

Commit:
Correct usage of done callback

- Switch to .includes for string containment check

ismay commented:
Ok, it took me a while to find the time to do this, but I've corrected my improper use of the `done` callback. I've also switched to the `.includes` method for the extension check, as per your suggestion @doodzik.

After looking at this again I do have to say that I think that there a quite some things I would like to do differently. I think a more functional approach would be nice (instead of mutating the files object directly, etc.). Plus some other patterns that just aren't very clear. Bit of a naive implementation, but then again, it's a starting point.

Also, maybe we could look at the way redux handles its async action creators for a way to implement our rendering function in a way that can be overridden by users (should they want to do so), and is more separated from this lib and layouts.

Anyway, I think this pr is complete. What do you think, ready for squash + merging?

doodzik commented:
this is a really good starting point. great work @ismay 👍 
can you elaborate on the async action creators?

woodyrew commented:
@ismay Not sure precisely what you mean by a more functional approach but it's better to mutate the `files` object with metalsmith in case a plugin is referencing any of the file objects. A similar situation to https://github.com/metalsmith/metalsmith/issues/247

ismay commented:
Ah ok, I didn't know that it was actually preferable to mutate, rather than return a clone. I was wondering whether cloning could create performance problems, but it seemed like a cleaner approach with less unintended side effects. Good to know that it's better to mutate in this case, thanks! 👍
